calico/felix/labels.py

```diff
@@ -0,0 +1,482 @@
+        stores the result.
+        """
+        _log.debug("Re-evaluating %s against %s (%s)", expr_id, item_id,
+                   label_values)
+        if expr is not None and label_values is not None:
+            now_matches = expr.evaluate(label_values)
+            _log.debug("After evaluation, now matches: %s", now_matches)
+        else:
+            _log.debug("Expr or labels missing: no match")
+            now_matches = False
+        # Update the index and generate events.  These methods are idempotent
+        # so they'll ignore duplicate updates.
+        if now_matches:
+            self._store_match(expr_id, item_id)
+        else:
+            self._discard_match(expr_id, item_id)
+
+    def _store_match(self, expr_id, item_id):
+        """
+        Stores that an expression matches an item.
+
+        Calls on_match_started() as a side-effect. Idempotent, does
+        nothing if the match is already recorded.
+        """
+        previously_matched = self.matches_by_expr_id.contains(expr_id, item_id)
+        if not previously_matched:
+            _log.debug("%s now matches: %s", expr_id, item_id)
+            self.matches_by_expr_id.add(expr_id, item_id)
+            self.matches_by_item_id.add(item_id, expr_id)
+            self.on_match_started(expr_id, item_id)
+
+    def _discard_match(self, expr_id, item_id):
+        """
+        Stores that an expression does not match an item.
+
+        Calls on_match_stopped() as a side-effect.  Idempotent, does
+        nothing if the non-match is already recorded.
+        """
+        previously_matched = self.matches_by_expr_id.contains(expr_id, item_id)
+        if previously_matched:
+            _log.debug("%s no-longer matches %s", expr_id, item_id)
+            self.matches_by_expr_id.discard(expr_id, item_id)
+            self.matches_by_item_id.discard(item_id, expr_id)
+            self.on_match_stopped(expr_id, item_id)
+
+    def on_match_started(self, expr_id, item_id):
+        """
+        Called when an expression starts matching a particular set of
+        labels.
+
+        Intended to be assigned/overriden.
+        """
+        _log.debug("SelectorExpression %s now matches item %s",
+                   expr_id, item_id)
+
+    def on_match_stopped(self, expr_id, item_id):
+        """
+        Called when an expression stops matching a particular set of
+        labels.
+
+        Intended to be assigned/overriden.
+        """
+        _log.debug("SelectorExpression %s no longer matches item %s",
+                   expr_id, item_id)
+
+
+class LabelValueIndex(LinearScanLabelIndex):
+    """
+    Label index that indexes the values of labels, allowing for efficient
+    (re)calculation of the matches for selectors of the form
+    'a == "b" && c == "d" && ...', which are the mainline.
+    """
+    def __init__(self):
+        super(LabelValueIndex, self).__init__()
+        self.item_ids_by_key_value = MultiDict()
```

Comment:
I've added a header comment to define the terms but I think the global rename would be quite a lot of work (given that "item" is used in a lot of places in the python APIs).  The intuition that I was working from was that labels are attached to items, so you have the ID of the items that the labels are for rather than the label_dict_id.